src/storage/s3.rs

```diff
@@ -0,0 +1,244 @@
+use super::Blob;
+use failure::Error;
+use futures::Future;
+use postgres::Connection;
+use rusoto_core::region::Region;
+use rusoto_credential::DefaultCredentialsProvider;
+use rusoto_s3::{GetObjectRequest, PutObjectRequest, S3Client, S3};
+use std::convert::TryInto;
+use std::io::Read;
+use time::Timespec;
+
+pub(crate) struct S3Backend<'a> {
+    client: S3Client,
+    bucket: &'a str,
+}
+
+impl<'a> S3Backend<'a> {
+    pub(crate) fn new(client: S3Client, bucket: &'a str) -> Self {
+        Self { client, bucket }
+    }
+
+    pub(super) fn get(&self, path: &str) -> Result<Blob, Error> {
+        let res = self
+            .client
+            .get_object(GetObjectRequest {
+                bucket: self.bucket.to_string(),
+                key: path.into(),
+                ..Default::default()
+            })
+            .sync()?;
+
+        let mut b = res.body.unwrap().into_blocking_read();
+        let mut content = Vec::with_capacity(
+            res.content_length
+                .and_then(|l| l.try_into().ok())
+                .unwrap_or(0),
+        );
+        b.read_to_end(&mut content).unwrap();
+
+        let date_updated = parse_timespec(&res.last_modified.unwrap())?;
+
+        Ok(Blob {
+            path: path.into(),
+            mime: res.content_type.unwrap(),
```

Comment:
Maybe mime normalization to `text/rust` and such should happen here if all downstream consumers use that

Reply:
I like having it closer where the mime is created so it's clear what's going on. If we normalized it just before uploading there would be some doubt whether code in the middle used the denormalized version.

Reply:
Well, isn't this the creation point of the `Blob`? I may be misunderstanding the code flow though

Reply:
Hmm, I misread the code too. This is when we fetch the file from S3, where it's already had a content-type set. The content-type should be correct in S3, if not it's a) impossible to fix manually and b) very unclear how things are happenin.

Reply:
Wouldn't S3 serve it back with the 'official' (and incorrect for what we need) mime types?

Reply:
No, we explicitly set the mimetype when uploading: https://github.com/rust-lang/docs.rs/pull/643/files#diff-c3a2a09475e9683a0e82b9a1d6f5e0b9R63

Reply:
Although in general I'm not super happy about setting an explicit content type, I think browser do a better job of mine detection than we could for source files.